// ngx-rangeslider-component/esm5/ngx-rangeslider-component.js
import { __spread } from 'tslib';
import { Component, ViewChild, ElementRef, HostListener, forwardRef, Renderer2, IterableDiffers, Input, Output, EventEmitter, Directive, NgModule } from '@angular/core';
import { NG_VALUE_ACCESSOR, NG_VALIDATORS } from '@angular/forms';
import { CommonModule } from '@angular/common';

var KEY_CODE = {
    RIGHT_ARROW: 39,
    LEFT_ARROW: 37,
};
KEY_CODE[KEY_CODE.RIGHT_ARROW] = 'RIGHT_ARROW';
KEY_CODE[KEY_CODE.LEFT_ARROW] = 'LEFT_ARROW';
var noop = function () { };
var RangeSliderComponent = /** @class */ (function () {
    function RangeSliderComponent(elementRef, renderer, iterableDiffers) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.iterableDiffers = iterableDiffers;
        this.minSliderClicked = false;
        this.maxSliderClicked = false;
        this.minSelected = false;
        this.maxSelected = false;
        this.toolTip = true;
        this.combineToolTip = false;
        this.minSliderTouched = false;
        this.onTouchedCallback = noop;
        this.onChangeCallback = noop;
        this.toolTips = [true, true];
        this.onRangeChange = new EventEmitter();
        this.iterableDiffer = this.iterableDiffers.find([]).create(null);
    }
    RangeSliderComponent.prototype.ngDoCheck = function () {
        var changes = this.iterableDiffer.diff(this.range);
        if (changes) {
            this.onChangeCallback(this.range);
            this.onTouchedCallback(this.range);
            this.onRangeChange.emit(this.range);
        }
    };
    RangeSliderComponent.prototype.onResize = function (event) {
        if (this.range) {
            this.getWidth();
        }
    };
    RangeSliderComponent.prototype.sethighlightClass = function (clas) {
        if (clas && this.sliderHilight) {
            this.renderer.removeClass(this.sliderHilight.nativeElement, 'dhighlightClass');
            this.renderer.addClass(this.sliderHilight.nativeElement, clas);
        }
    };
    RangeSliderComponent.prototype.setBarClass = function (clas) {
        if (clas && this.bar) {
            this.renderer.removeClass(this.bar.nativeElement, 'dbarClass');
            this.renderer.addClass(this.bar.nativeElement, clas);
        }
    };
    RangeSliderComponent.prototype.setSliderClass = function (clas) {
        if (clas && this.minSlider && this.maxSlider) {
            this.renderer.removeClass(this.minSlider.nativeElement, 'dsliderClass');
            this.renderer.removeClass(this.maxSlider.nativeElement, 'dsliderClass');
            this.renderer.addClass(this.minSlider.nativeElement, clas);
            this.renderer.addClass(this.maxSlider.nativeElement, clas);
        }
    };
    RangeSliderComponent.prototype.writeValue = function (value) {
        if (value) {
            if (value[0] === null) {
                value[0] = this.min;
            }
            else if (value[1] === null) {
                value[1] = this.max;
            }
            else {
                if (this.range) {
                    var prevRange = this.rangeCache.slice(0);
                    if (value[0] > prevRange[1]) {
                        value[0] = prevRange[1];
                    }
                    else if (value[0] < this.min) {
                        value[0] = this.min;
                    }
                    else if (value[1] < prevRange[0]) {
                        value[1] = prevRange[0];
                    }
                    else if (value[1] > this.max) {
                        value[1] = this.max;
                    }
                }
            }
            this.update(value);
        }
    };
    RangeSliderComponent.prototype.update = function (range) {
        this.range = __spread(range);
        this.rangeCache = (JSON.parse(JSON.stringify(range)));
        this.setBarClass(this.barClass);
        this.setSliderClass(this.sliderClass);
        this.sethighlightClass(this.highlightClass);
        this.getWidth();
    };
    RangeSliderComponent.prototype.minTouched = function (event) {
        var evt = event.changedTouches[0];
        this.minMouseDown(evt);
    };
    RangeSliderComponent.prototype.maxTouched = function (event) {
        var evt = event.changedTouches[0];
        this.maxMouseDown(evt);
    };
    RangeSliderComponent.prototype.minMouseDown = function (event) {
        this.minSliderClicked = true;
        this.minSelected = true;
        this.maxSliderClicked = false;
        this.maxSelected = false;
        this.minSliderIntialLeft = event.target.offsetLeft;
        this.intialMinMouseX = event.clientX;
    };
    RangeSliderComponent.prototype.maxMouseDown = function (event) {
        this.maxSliderClicked = true;
        this.maxSelected = true;
        this.minSliderClicked = false;
        this.minSelected = false;
        this.maxSliderIntialLeft = event.target.offsetLeft;
        this.intialMaxMouseX = event.clientX;
    };
    RangeSliderComponent.prototype.touchMove = function (event) {
        var evt = event.changedTouches[0];
        this.mouseMove(evt);
    };
    RangeSliderComponent.prototype.registerOnChange = function (onChange) {
        this.onChangeCallback = onChange;
    };
    RangeSliderComponent.prototype.registerOnTouched = function (onTouched) {
        this.registerOnTouched = onTouched;
    };
    RangeSliderComponent.prototype.validate = function () {
        return null;
    };
    RangeSliderComponent.prototype.getlength = function (num) {
        return String(num).match(/\d/g).length;
    };
    RangeSliderComponent.prototype.getWidth = function () {
        if (this.bar && this.minSlider && this.range && this.range[0] !== undefined) {
            this.sliderWidth = this.minSlider.nativeElement.offsetWidth;
            this.sliderHeight = this.minSlider.nativeElement.offsetHeight;
            this.toolTipTop = (this.sliderHeight + 10) * -1;
            this.barWidth = this.bar.nativeElement.offsetWidth;
            if (this.sliderWidth && this.barWidth) {
                this.rangeDiff = this.max - this.min;
                this.rangeInPixels = this.barWidth - this.sliderWidth;
                if (this.barWidth && this.sliderWidth) {
                    this.valToPixelFactor = ((this.rangeInPixels) / this.rangeDiff);
                }
                this.minSliderLeft = (this.range[0] - this.min) * this.valToPixelFactor;
                this.maxSliderLeft = (this.range[1] - this.min) * this.valToPixelFactor;
            }
            this.highlightDimensions();
        }
    };
    RangeSliderComponent.prototype.highlightDimensions = function () {
        this.highlightLeft = this.minSliderLeft + (this.sliderWidth / 2);
        this.highlightWidth = this.maxSliderLeft - this.minSliderLeft;
        if (this.range[0] !== undefined) {
            this.minToolTipWidth = this.getlength(this.range[0].toString());
            var cond = this.minToolTipWidth * 8 + this.minSliderLeft + 8;
            if (cond > this.maxSliderLeft && this.toolTips[0] && this.toolTips[1]) {
                this.toolTip = false;
                this.combineToolTip = true;
                this.combineToolTipWidth = this.getlength(this.range[0] + "-" + this.range[1]) * 8;
                var maxLeft = this.rangeInPixels - this.combineToolTipWidth;
                this.combineToolTipLeft = this.minSliderLeft < maxLeft ? this.minSliderLeft : maxLeft;
            }
            else {
                this.toolTip = true;
                this.combineToolTip = false;
            }
        }
    };
    RangeSliderComponent.prototype.clickedOnBar = function (event) {
        this.minSelected = false;
        this.maxSelected = false;
        var targetId = event.target.id;
        var left = event.offsetX;
        if (targetId === 'bar') {
            if (left < this.minSliderLeft) {
                var value = this.pixToVal(this.min, left);
                var finalVal = this.stepr(value);
                this.range[0] = finalVal <= this.min ? this.min : finalVal;
                this.minSliderLeft = this.valToPixel(this.range[0]);
            }
            else if (left > this.maxSliderLeft) {
                if (left >= (this.rangeInPixels)) {
                    this.maxSliderLeft = this.rangeInPixels;
                    this.range[1] = this.max;
                }
                else {
                    var value = this.pixToVal(this.min, left);
                    this.range[1] = this.stepr(value);
                    this.maxSliderLeft = this.valToPixel(this.range[1]);
                }
            }
        }
        this.highlightDimensions();
    };
    RangeSliderComponent.prototype.clickedOnhighlight = function (event) {
        this.minSelected = false;
        this.maxSelected = false;
        var left = event.offsetX;
        var cond = (this.highlightWidth / 2);
        if (left <= cond) {
            var orgLeft = this.minSliderLeft + (this.sliderWidth / 2) + left;
            var value = this.pixToVal(this.min, orgLeft);
            this.range[0] = this.stepr(value);
            this.minSliderLeft = this.valToPixel(this.range[0]);
        }
        else if (left > cond) {
            var orgLeft = this.minSliderLeft + (this.sliderWidth / 2) + left;
            var value = this.pixToVal(this.min, orgLeft);
            this.range[1] = this.stepr(value);
            this.maxSliderLeft = this.valToPixel(this.range[1]);
        }
        this.highlightDimensions();
    };
    RangeSliderComponent.prototype.pixToVal = function (min, left) {
        var value = Number((min + left * (1 / this.valToPixelFactor)).toFixed(2));
        return value;
    };
    RangeSliderComponent.prototype.stepr = function (value) {
        if (this.step) {
            var fin = value - Math.floor(value);
            if (fin >= 0.5) {
                value = Math.ceil(value);
            }
            else {
                value = Math.floor(value);
            }
            var remainder = value % this.step;
            if (remainder === 0) {
                return value;
            }
            else {
                if (remainder >= (this.step / 2)) {
                    value = value + (this.step - remainder);
                }
                else {
                    value = value - remainder;
                }
            }
        }
        return value;
    };
    RangeSliderComponent.prototype.valToPixel = function (value) {
        var pixel = this.valToPixelFactor * (value - this.min);
        return pixel;
    };
    RangeSliderComponent.prototype.mouseMove = function (event) {
        if (this.minSelected || this.maxSelected) {
            if (this.minSelected) {
                this.minChange = event.clientX - this.intialMinMouseX;
                var left = this.minSliderIntialLeft + this.minChange;
                var value = this.pixToVal(this.min, left);
                if (value <= this.range[1]) {
                    if (value <= this.min) {
                        this.minSliderLeft = 0;
                        this.range[0] = this.min;
                    }
                    else {
                        var finalVal = this.stepr(value);
                        this.range[0] = finalVal <= this.range[1] ? finalVal : this.range[1];
                        this.minSliderLeft = this.valToPixel(this.range[0]);
                    }
                }
            }
            else if (this.maxSelected) {
                this.maxChange = event.clientX - this.intialMaxMouseX;
                var left = this.maxSliderIntialLeft + this.maxChange;
                var value = this.pixToVal(this.min, left);
                if (value >= this.range[0]) {
                    if (value >= this.max) {
                        this.maxSliderLeft = this.rangeInPixels;
                        this.range[1] = this.max;
                    }
                    else {
                        var final = this.stepr(value);
                        this.range[1] = final > this.range[0] ? final <= this.max ? final : this.max : this.range[0];
                        this.maxSliderLeft = this.valToPixel(this.range[1]);
                    }
                }
            }
            this.highlightDimensions();
        }
        return false;
    };
    RangeSliderComponent.prototype.mouseUp = function (event) {
        this.minSelected = false;
        this.maxSelected = false;
    };
    RangeSliderComponent.prototype.clickedOutside = function (event) {
        if (event) {
            this.minSliderClicked = false;
            this.maxSliderClicked = false;
        }
    };
    RangeSliderComponent.prototype.onKeyDown = function (event) {
        var step = this.step ? this.step : 1;
        if (this.minSliderClicked || this.maxSliderClicked) {
            if (this.minSliderClicked) {
                if (event.keyCode === KEY_CODE.LEFT_ARROW) {
                    var left = this.minSliderLeft - (step * this.valToPixelFactor);
                    var value = this.pixToVal(this.min, left);
                    if (value >= this.min) {
                        this.range[0] = this.stepr(value);
                        this.minSliderLeft = this.valToPixel(this.range[0]);
                        this.minSliderIntialLeft = this.minSliderLeft;
                    }
                }
                else if (event.keyCode === KEY_CODE.RIGHT_ARROW) {
                    var left = this.minSliderLeft + (step * this.valToPixelFactor);
                    var value = this.pixToVal(this.min, left);
                    if (value <= this.range[1]) {
                        this.range[0] = this.stepr(value);
                        this.minSliderLeft = this.valToPixel(this.range[0]);
                        this.minSliderIntialLeft = this.minSliderLeft;
                    }
                }
            }
            else if (this.maxSliderClicked) {
                if (event.keyCode === KEY_CODE.LEFT_ARROW) {
                    var left = this.maxSliderLeft - (step * this.valToPixelFactor);
                    var value = this.pixToVal(this.min, left);
                    if (value >= this.range[0]) {
                        this.range[1] = this.stepr(value);
                        this.maxSliderLeft = this.valToPixel(this.range[1]);
                        this.maxSliderIntialLeft = this.maxSliderLeft;
                    }
                }
                else if (event.keyCode === KEY_CODE.RIGHT_ARROW) {
                    var left = this.maxSliderLeft + (step * this.valToPixelFactor);
                    var value = this.pixToVal(this.min, left);
                    if (value <= this.max) {
                        this.range[1] = this.stepr(value);
                        this.maxSliderLeft = this.valToPixel(this.range[1]);
                        this.maxSliderIntialLeft = this.maxSliderLeft;
                    }
                }
            }
            this.highlightDimensions();
        }
    };
    return RangeSliderComponent;
}());
RangeSliderComponent.decorators = [
    { type: Component, args: [{
                selector: 'range-slider',
                template: "\n<div (clickOutside)=\"clickedOutside($event)\" id=\"bar\"  #bar   class=\"slidersContainer dbarClass\" \n  (click)=\"clickedOnBar($event)\"  (touchmove)=\"touchMove($event)\"\n  (touchend)=\"mouseUp($event)\">\n\n      \n\n      <span style=\"position:absolute\" *ngIf=\"range && toolTip && toolTips[0]\"  [style.top.px]=\"toolTipTop\" [style.left.px]=\"minSliderLeft\" >{{range[0]}}</span>\n      <span style=\"position:absolute\" *ngIf=\"range && toolTip && toolTips[1]\" [style.top.px]=\"toolTipTop\" [style.left.px]=\"maxSliderLeft\" >{{range[1]}}</span>\n      <span style=\"position:absolute\" *ngIf=\"range && combineToolTip\" [style.top.px]=\"toolTipTop\" [style.left.px]=\"combineToolTipLeft\"  >{{range[0]}}-{{range[1]}}</span> \n\n  <div #minSlider   class=\"dslider minSlider dsliderClass\" (mousedown)=\"minMouseDown($event)\"  \n  [style.left.px]=\"minSliderLeft\" (mouseup)=\"mouseUp($event)\" (touchstart)=\"minTouched($event)\">\n  \n</div>\n\n  <div #maxSlider class=\"dslider maxSlider dsliderClass\" (mousedown)=\"maxMouseDown($event)\" \n   [style.left.px]=\"maxSliderLeft\" (mouseup)=\"mouseUp($event)\" (touchstart)=\"maxTouched($event)\">\n   \n  </div>\n\n  <div  #sliderHilight class=\"sliderHilight dhighlightClass\" (click)=\"clickedOnhighlight($event)\"\n  [style.width.px]=\"highlightWidth\" [style.left.px]=\"highlightLeft\" ></div>\n  </div>",
                styles: ["div{-webkit-box-sizing:border-box;box-sizing:border-box}.slidersContainer{position:relative;margin:20px 0}.dslider{position:absolute;cursor:pointer;bottom:-50%}.minSlider{z-index:1010}.maxSlider{z-index:1015}.sliderHilight{position:absolute;z-index:1005;height:inherit}.dbarClass{width:100%;height:10px;background:#d3d3d3}.dsliderClass{width:20px;height:20px;background:#fff;border:1px solid #4169e1}.dhighlightClass{background:#4169e1}.tooltipText{text-align:center;padding:5px 0;position:absolute;bottom:72%;left:13%}"],
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(function () { return RangeSliderComponent; }),
                        multi: true
                    },
                    { provide: NG_VALIDATORS, useExisting: forwardRef(function () { return RangeSliderComponent; }), multi: true }
                ]
            },] },
];
RangeSliderComponent.ctorParameters = function () { return [
    { type: ElementRef },
    { type: Renderer2 },
    { type: IterableDiffers }
]; };
RangeSliderComponent.propDecorators = {
    min: [{ type: Input }],
    max: [{ type: Input }],
    toolTips: [{ type: Input }],
    step: [{ type: Input }],
    onRangeChange: [{ type: Output }],
    bar: [{ type: ViewChild, args: ['bar',] }],
    minSlider: [{ type: ViewChild, args: ['minSlider',] }],
    maxSlider: [{ type: ViewChild, args: ['maxSlider',] }],
    sliderHilight: [{ type: ViewChild, args: ['sliderHilight',] }],
    onResize: [{ type: HostListener, args: ['window:resize', ['$event'],] }],
    highlightClass: [{ type: Input }],
    barClass: [{ type: Input }],
    sliderClass: [{ type: Input }],
    mouseMove: [{ type: HostListener, args: ['window:mousemove', ['$event'],] }],
    mouseUp: [{ type: HostListener, args: ['window:mouseup', ['$event'],] }],
    onKeyDown: [{ type: HostListener, args: ['window:keydown', ['$event'],] }]
};
var ClickOutsideDirective = /** @class */ (function () {
    function ClickOutsideDirective(elementRef) {
        this.elementRef = elementRef;
        this.clickOutside = new EventEmitter();
    }
    ClickOutsideDirective.prototype.onClick = function (targetElement) {
        var clickedInside = this.elementRef.nativeElement.contains(targetElement);
        if (!clickedInside) {
            this.clickOutside.emit(true);
        }
    };
    return ClickOutsideDirective;
}());
ClickOutsideDirective.decorators = [
    { type: Directive, args: [{
                selector: '[clickOutside]'
            },] },
];
ClickOutsideDirective.ctorParameters = function () { return [
    { type: ElementRef }
]; };
ClickOutsideDirective.propDecorators = {
    clickOutside: [{ type: Output }],
    onClick: [{ type: HostListener, args: ['click', ['$event.target'],] }]
};
var RangeSliderModule = /** @class */ (function () {
    function RangeSliderModule() {
    }
    return RangeSliderModule;
}());
RangeSliderModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [RangeSliderComponent, ClickOutsideDirective],
                exports: [RangeSliderComponent, ClickOutsideDirective]
            },] },
];

export { RangeSliderModule, ClickOutsideDirective as ɵb, RangeSliderComponent as ɵa };
